### PR TITLE
BFD-1911: Add synthea validation script

### PR DESCRIPTION
**Pull request by lsmitchell**
**JIRA Ticket:**
[BFD-1911](https://jira.cms.gov/browse/BFD-1911)

**User Story or Bug Summary:**

As a BFD Engineer I would like to know that the synthetic data I am loading will not conflict with existing data so I can avoid loading errors and wasted time remediating failed loads.

---

### What Does This PR Do?

- Adds a script that validates a target synthea end state properties file against a target DB to verify there will be no conflicts with the inserted data on the incremented fields with constrained unique values
    - This is intended to be used in future synthea generation automation
    - This script will be enhanced in the future to also update the synthea properties file, and has some stubbed functionality to add that piece
- Fixes the script that adjusted the clm_grp_id ranges to use threading to speed it up (this was the script used in test, just checking it in now)

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by mackec1 — DISMISSED** [review]
Looks reasonable; I might have created a failure mask to discern which claims table(s) are in error, but I'm supposing that if one fails, the whole thing is a failure. Then it becomes a case for further digging to discern which claims tables are good vs. bad?

File: ops/ccs-ops-misc/synthetic-data/scripts/prepare-for-new-synthea-load.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    
+    fi_num_start = re.findall("exporter.bfd.fi_doc_cntl_num_start.*$",properties_file,re.MULTILINE)[0].split("=")[1]
+    query = f"select count(*) from hospice_claims where fi_doc_clm_cntl_num::bigint <= {fi_num_start}"
+    result = result + _execute_single_count_query(db_string, query)
+    
+    if result > 0:
+        print(f"fi_doc_cntl_num id range invalid, {result} values past starting id {fi_num_start} across all tables")
+        ranges_good = False
+    else:
+        print("fi_doc_cntl_num range is valid")
```
Comment (christrotteradhoc) [inline ops/ccs-ops-misc/synthetic-data/scripts/prepare-for-new-synthea-load.py:136]: Why not have this check and explicit message for each of the claim types like what you have for carrier, dme, and benes?

Comment (christrotteradhoc) [reply ops/ccs-ops-misc/synthetic-data/scripts/prepare-for-new-synthea-load.py:136]: possibly even a function to abstract this

Comment (lsmitchell) [reply ops/ccs-ops-misc/synthetic-data/scripts/prepare-for-new-synthea-load.py]: Good call; I've updated the output to be more precise, abstracted the logic, and added it for all the tables/fields checked individually.

File: ops/ccs-ops-misc/synthetic-data/scripts/fix-legacy-clm-grp.py [hunk 2]
```diff
@@ -90,51 +88,86 @@ def replace_grp_ids(table_ids, mapped_lists, db_string):
...
     """
+    
+    process = Process(target=replace_ids, args=(table_ids["carrier_grp_ids"], mapped_lists["carrier_grp_ids"], "carrier_claims_new", db_string,))
+    process.start()
+    
+    process2 = Process(target=replace_ids, args=(table_ids["inpatient_grp_ids"], mapped_lists["inpatient_grp_ids"], "inpatient_claims_new", db_string,))
+    process2.start()
+    
+    process3 = Process(target=replace_ids, args=(table_ids["outpatient_grp_ids"], mapped_lists["outpatient_grp_ids"], "outpatient_claims_new", db_string,))
+    process3.start()
```
[the remark this answers is not in the document]
Comment (lsmitchell) [reply ops/ccs-ops-misc/synthetic-data/scripts/fix-legacy-clm-grp.py]: Thanks for noticing this, fixed!

**Comment by lsmitchell** [comment]
Some output examples with the new more precise output:
Full validation success
image
Full validation failure:
image
Partial failure (with success at the end to prove the rolling overall validation check works):
image

**Review by christrotteradhoc — APPROVED** [review]

**Review by mackec1 — APPROVED** [review]

**Pull request merged**